age-piv-p256.md

```diff
@@ -0,0 +1,100 @@
+# age piv-p256 recipient stanza
+
+[c2sp.org/age-piv-p256](https://c2sp.org/age-piv-p256)
+
+## Introduction
+
+This document specifies the age piv-p256 recipient stanza. This stanza
+is used for encryption using NIST P-256 keys stored on hardware tokens
+or PIV devices.
+
+Apart from using the NIST P-256 curve, the piv-p256 recipient stanza
+differs from the native [X25519 recipient
+stanza](https://age-encryption.org/v1#x25519-recipient-stanza) in that
+it includes a tag of the recipient, to allow the recipient to select the
+correct secret without the hardware token requesting user interaction.
+
+## Conventions used in this document
+
+ABNF syntax follows [RFC
+5234](https://www.rfc-editor.org/rfc/rfc5234.html) and [RFC
+7405](https://www.rfc-editor.org/rfc/rfc7405.html) and references the
+core rules in RFC 5234, Appendix B.1.
+
+The base64 encoding used throughout is the standard Base 64 encoding
+specified in [RFC 4648](https://www.rfc-editor.org/rfc/rfc4648.html),
+Section 4, without `=` padding characters (sometimes referred to as
+"raw" or "unpadded" base64). Encoders MUST generate canonical base64
+according to RFC 4648, Section 3.5, and decoders MUST reject
+non-canonical encodings and encodings ending with `=` padding
+characters.
+
+`||` denotes concatenation. `0x` followed by two hexadecimal characters
+denotes a byte value in the 0-255 range.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" in this document are to be interpreted as described in [BCP
+14](https://www.rfc-editor.org/info/bcp14) [RFC
+2119](https://www.rfc-editor.org/rfc/rfc2119.html) [RFC
+8174](https://www.rfc-editor.org/rfc/rfc8174.html) when, and only when,
+they appear in all capitals, as shown here.
+
+## Recipient stanza
+
+A piv-p256 recipient stanza has three arguments.
```

Comment:
If we make other changes, we should drop "piv" from the name.

Maybe tagged-p256? Or hw-p256?

I assume @Foxboron could use it for age-plugin-tpm, too.

I wonder if we should standardize the recipient (not identity) format, and add native support to age(1) so that one doesn't need the plugin to encrypt to PIV/TPM/SE/any other P-256 identity.

/cc @str4d

Reply:
I can probably use this, I wasn't aware of this PR work when I originally wrote `age-plugin-tpm`.

Reply:
Native support sounds neat.

- You don't need to install a plugin to encrypt to these plugins. For encrypting to a different person, it feels odd that you need to install `age-plugin-se` even if you don't have an Apple device. It also avoids the plugin discovery problem where you need to explain which plugin people have to download to encrypt a message to you.
- For `age-plugin-se`, I could drop all cross-platform support, since the parts that don't involve encryption are only useful on macOS. I can imagine there are other hardware encryption plugins that don't make sense on all platforms.
- It hides information about which type of token corresponds to a certain recipient
- If there's a backwards-incompatible change now, only half of the existing state machines need to be updated (the easy half)